drivers/network/tcpip/tcpip/icmp.c

```diff
@@ -0,0 +1,389 @@
+        {
+            UINT32 _DataSize = min(Context->RemainingSize, DataSize);
+
+            memcpy(Context->CurrentReply + Context->RemainingSize - _DataSize, DataBuffer, _DataSize);
+            CurrentReply->Data = CurrentUserBuffer + Context->RemainingSize - _DataSize;
+            CurrentReply->DataSize = _DataSize;
+
+            Context->RemainingSize -= _DataSize;
+            // Context->ReplyBuffer += _DataSize;
+        }
+        else
+        {
+            TI_DbgPrint(DEBUG_ICMP, ("RemainingSize: %u, DataSize: %d\n", Context->RemainingSize, DataSize));
+        }
+
+        if (OptionsLength > 0 && Context->RemainingSize > 0)
+        {
+            UINT32 _OptSize = min(Context->RemainingSize, OptionsLength);
+
+            memcpy(Context->CurrentReply + Context->RemainingSize + _OptSize, Options, _OptSize);
+            CurrentReply->Options.OptionsData = CurrentUserBuffer + Context->RemainingSize + _OptSize;
+            CurrentReply->Options.OptionsSize = _OptSize;
+
+            Context->RemainingSize -= _OptSize;
+            // Context->ReplyBuffer += _OptSize;
+        }
+        else
+        {
+            TI_DbgPrint(DEBUG_ICMP, ("RemainingSize: %u, OptSize: %d\n", Context->RemainingSize, OptionsLength));
+        }
+
+        Context->nReplies++;
+    }
+
+    if (NT_ERROR(Context->ErrorStatus) || Context->RemainingSize < sizeof(ICMP_ECHO_REPLY))
+    {
+        TI_DbgPrint(DEBUG_ICMP, ("The space is over: %u, or error happened: 0x%x\n",
+            Context->RemainingSize, Context->ErrorStatus));
+
+        // if the timer was inserted, that means DPC has not been queued yet
+        if (KeCancelTimer(&TimeoutTimer))
+        {
+            ASSERT(KeInsertQueueDpc(&TimeoutDpc, NULL, NULL));
+        }
+    }
+
+    KeReleaseSpinLock(&Context->ProcessingLock, OldIrql);
+    return STATUS_SUCCESS;
+}
+
+KDEFERRED_ROUTINE EndRequestHandler;
+
+VOID
+NTAPI
+EndRequestHandler(
+    _In_ PKDPC Dpc,
+    _In_opt_ PVOID _Context,
+    _In_opt_ PVOID SystemArgument1,
+    _In_opt_ PVOID SystemArgument2)
+{
+    PICMP_PACKET_CONTEXT Context = (PICMP_PACKET_CONTEXT)_Context;
+    PIO_STACK_LOCATION CurrentStack;
+    UINT32 ReplyLength;
+    PIRP Irp;
+
+    TI_DbgPrint(DEBUG_ICMP, ("Finishing request Context: %p\n", Context));
+
+    KeAcquireSpinLockAtDpcLevel(&Context->ProcessingLock);
+
+    // first: stop receiving new packets
+    {
+        NTSTATUS _Status = FileCloseAddress(&Context->TdiRequest);
+        ASSERT(NT_SUCCESS(_Status));
+    }
+
+    Irp = Context->Irp;
+    CurrentStack = IoGetCurrentIrpStackLocation(Irp);
+    ReplyLength = CurrentStack->Parameters.DeviceIoControl.OutputBufferLength;
+
+    // second: prepare return data
+    if (Context->nReplies > 0)
+    {
+        ((PICMP_ECHO_REPLY)Irp->AssociatedIrp.SystemBuffer)->Reserved = Context->nReplies;
+        Irp->IoStatus.Status = Context->ErrorStatus;
+        Irp->IoStatus.Information = ReplyLength;
```

Comment:
The `SystemBuffer` is a pool allocation that's not initialized to zero. Since you're always returning the full size, you may have uninitialized data at the end (or given how the buffer is filled, in the middle).
So you either need a different allocation scheme for data in the buffer, or it needs to be zeroed before output is written to it.